crates/holochain_zome_types/src/timestamp.rs

```diff
@@ -1,16 +1,398 @@
+/// - Field 1: u32 - Nanoseconds in addition to above seconds, always in positive direction.
+///
+/// Supports +/- chrono::Duration directly.  There is no Timestamp::now() method, since this is not
+/// supported by WASM; however, holochain_types provides a timestamp::now() method.
+///
+/// Create a new Timestamp instance from the supplied secs/nsecs.  Note that we can easily create a
+/// Timestamp that cannot be converted to a valid DateTime<Utc> (ie. by supplying 86,400-second days
+/// beyond range of +/- i32 offset from 0AD or 1970AD, nsecs beyond 1e9, etc.; see its code.)
+#[derive(
+    Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Deserialize, Serialize, SerializedBytes,
+)]
 pub struct Timestamp(
-    // sec
-    pub i64,
-    // nsec
-    pub u32,
+    pub i64, // seconds from UNIX Epoch, positive or negative
+    pub u32, // nanoseconds, always a positive offset
 );
+
+/// Display as RFC3339 Date+Time for sane value ranges (0000-9999AD).  Beyond that, format
+/// as (seconds, nanoseconds) tuple (output and parsing of large +/- years is unreliable).
+impl fmt::Display for Timestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+	let ce = -62167219200_i64..=253402214400_i64;
+	if ce.contains(&self.0) {
+	    if let Ok(ts) = chrono::DateTime::<chrono::Utc>::try_from(self) {
+		return write!(f, "{}", ts.to_rfc3339_opts(chrono::SecondsFormat::AutoSi, true));
+	    }
+	}
+	// Outside 0000-01-01 to 9999-12-31; Display raw value tuple, or not a valid DateTime<Utc>;
+	// Display raw value tuple
+	write!(f, "({},{})", self.0, self.1)
+    }
+}
+
+impl fmt::Debug for Timestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Timestamp({})", self)
+    }
+}
+
+/// Infallible conversions into a Timestamp.  The only infallible ways to create a Timestamp are
+/// `from` a Unix timestamp, or `normalize` with a timestamp and nanoseconds, or converting from
+/// a DateTime<Utc>.
+impl From<i64> for Timestamp {
+    fn from(secs: i64) -> Self {
+        Self(secs, 0)
+    }
+}
+
+impl From<i32> for Timestamp {
+    fn from(secs: i32) -> Self {
+        Self(secs.into(), 0)
+    }
+}
+
+impl From<u32> for Timestamp {
+    fn from(secs: u32) -> Self {
+        Self(secs.into(), 0)
+    }
+}
+
+impl From<chrono::DateTime<chrono::Utc>> for Timestamp {
+    fn from(t: chrono::DateTime<chrono::Utc>) -> Self {
+        std::convert::From::from(&t)
+    }
+}
+
+impl From<&chrono::DateTime<chrono::Utc>> for Timestamp {
+    fn from(t: &chrono::DateTime<chrono::Utc>) -> Self {
+        let t = t.naive_utc();
+        Timestamp(t.timestamp(), t.timestamp_subsec_nanos())
+    }
+}
+
+// There is *no* infallible conversion from a Timestamp to a DateTime.  These may panic in
+// from_timestamp due to out-of-range secs or nsecs, making all code using/displaying a Timestamp
+// this way dangerously fragile!
+/*
+impl From<Timestamp> for chrono::DateTime<chrono::Utc> {
```

Comment:
if the intent here is to scare people away from implementing this without realising, i doubt they will ever read this comment

what about something like this?

```rust
impl From<Timestamp> for chrono::DateTime<chrono::Utc> {
  panic!("Converting from a timestamp to a DateTime without error handling is always a bug. Use TryFrom instead.");
}
```

Reply:
I think its best just to not provide the infallible conversions; I adjusted the comment to be an implementation note to someone who happens to look at the code to see why there aren't any infallible conversion.